### DISCUSSION TITLE
Mujoco Collision Detection vs other Calculations

### DISCUSSION THREAD

**Discussion by felwoy** [google-deepmind/mujoco#337] [Asking for Help] [answered]
Hello,

so I am not very experienced in using MuJoCo, but i am building and experementing with a Drone/Quadcopter in MuJoCo. I know it's not the main intention behind MuJoCo, but still want to try.
So I understood that for collision detection MuJoCo uses Covex Hulls if not provided proper convex files. When building a rotor, the rotor itself, determined by it's shape, generates lift/lifting force. Since the rotor shape does or should matter for that calculation, I am interested in whether the underlying stl-shape is used for calculating lift or the convex hull of it body? So I can decide whether to split up the parts to get better/convexer bodies or not.

Also if someone has knowledge about similar attempts or thoughts I'd be interested in a short opinion in how bad of an idea it is to build drones/fluid dynamics simulations in MuJoCo and maybe some reasons and things that i should be aware of.

Thank you for any kind of help, appreciate it.

**Comment by rohit-kumar-j** [comment]
@felwoy Mujoco has a basic [fluid dynamics model](https://mujoco.readthedocs.io/en/latest/computation.html?highlight=fluid#passive-forces) to replicate some of the macro effects. If you are looking for drag forces and vortex formations, I don't think that it would be possible. From docs:

> Proper simulation of fluid dynamics is beyond the scope of MuJoCo, and would be too slow for the applications we aim to facilitate. Nevertheless we provide a phenomenological model which is sufficient for simulating behaviors such as flying and swimming. It is enabled by setting `mjModel.opt.viscosity` and `mjModel.opt.density` to positive values (they are zero by default.)

Seems like the rotor shape doesn't matter:

> The shape of each body for fluid dynamics purposes is assumed to be the equivalent inertia box, which can also be visualized. Each forward-facing (relative to the linear velocity) face of the box experiences force along its normal direction. All faces also experience torque due to the angular velocity; this torque is obtained by integrating the force resulting from the rotation over the surface area. 

> Existing phenomenological models of fluid dynamics tend to be valid for one regime (e.g. flat vs. spherical objects) and it is difficult to construct a model which can be simulated efficiently and is broadly valid. For example, drag forces are known to transition from being linear in velocity at low Reynolds numbers, to being quadratic in velocity at high Reynolds numbers. However it is not clear how this transition should occur for lift forces. This motivated the above separation, using density to compute all quadratic forces and viscosity to compute all linear forces. Users who need a more detailed simulation of fluid dynamics should leave the density and viscosity parameters set to zero so as to disable the built-in mechanism. A custom model of fluid dynamics (or any other force field that depends only on position and velocity) can be implemented in the callback [mjcb_passive](https://mujoco.readthedocs.io/en/latest/APIreference.html#mjcb-passive).

**Reply by felwoy** [reply]
Thank you very much.
So if I understand inertia box correctly, it would make a difference and actually an important one if I used one .stl
 for the rotor or cut it in multiple files for better inertia boxes what is kinda similar to cutting for convex hulls.
![rotor](https://user-images.githubusercontent.com/91830402/173027753-2f643273-65f4-4abb-8761-02a705358a0c.png)
Also it is stated that inertia boxes can be visualized. How could i activate it, in the python script or in the xml-file would be working best for me.

**Reply by rohit-kumar-j** [reply]
I don't think it is wise to cut the rotor into multiple geoms or bodies. The actual aerodynamics are not going to be taken into consideration or will be accurate. You could replicate/try to emulate the forces with something like [gym-pybullet-drones](https://github.com/utiasDSL/gym-pybullet-drones). It mimics the effects of drones pretty well instead of overloading on the computation.

**Reply by rohit-kumar-j** [reply]
For visualizing inertia in mujoco with python, you could use [mujoco-python-viewer](https://github.com/rohanpsingh/mujoco-python-viewer).

**Comment by yuvaltassa** [comment, marked as answer]
1. First, MuJoCo now includes a much more advanced fluid dynamics model mentioned in item 15 [here](https://mujoco.readthedocs.io/en/latest/changelog.html#id7). However it is currently an experimental feature and not very well documented. We expect this feature to come out of experimental status over the summer.
2. However, despite being much more advanced than the default model, @rohit-kumar-j is correct that this is likely not right way to model a propellor, though it is certainly possible.
3. The recommended way to model drone propellors is by using `site` actuation targets, with no degrees-of-freedom given to propellor rotation. These are described [here](https://mujoco.readthedocs.io/en/latest/XMLreference.html#actuator-general). NB that site actuators respect the full 6D `gear` specification, so you will likely want to place a site at the centre of the propellor (say with its z-axis pointing up) and use `gear="0 0 a 0 0 b"`, where `a` denotes the coefficient of force applied along the z-axis and `b` denotes the coefficient of torque along the z-axis. The sign of `b` will be determined by the direction of propellor spin.

Please let us know how you get on with your project, we are looking to add such a drone model to our library of example models.

**Reply by felwoy** [reply]
1. This helped a lot. So by presenting the fluidshape attribute to a geom fluid dynamic calculations will be done based on shape rather than on inertia boxes. So using the rotor.stl-file in mesh-type geom would be the most accurate way to represent the rotor.
3. Right now i am just using hinge joints and motor actuators but i will certainly try this later on. Thank you very much.

I will, since this is my first bigger mujoco-project i will certainly have follow up questions ;)
Would be amazing to have such a model. I started project planing phase with elaborating mujoco capabilities of simulating flying/lifting, because I wasn't aware of anything that would proof that something like this certainly would work.

**Comment by robotlearning123** [comment]
Is there any plans for more underwater robot demos? Thank you. @yuvaltassa

**Reply by yuvaltassa** [reply]
Hi @wangcongrobot,

We would love to add an underwater robot example! Can you find a specific robot that you would like us to implement as an example? It is critical that the model is provided under a permissive license so that we can redistribute.

I would also prefer a bio-mimetic one with fins rather than propellors, if possible.

**Reply by robotlearning123** [reply]
Hi @yuvaltassa, thanks. There are some projects/papers related to bio-mimetic simulation:

- [gym-fish](https://github.com/fish-gym/gym-fish)
- https://ieeexplore.ieee.org/document/9561145

I would like to help with the example if possible. Thank you.

**Reply by robotlearning123** [reply]
This video looks interesting.
https://youtu.be/fBoAKVq2Tgo

**Reply by yuvaltassa** [reply]
Yup these all look interesting. As a side note, you know that the DM Control Suite has two [swimming models right](https://www.youtube.com/watch?v=rAai4QzcYbs&t=38s)?

These models use MuJoCo's old fluid model rather than the new one unfortunately are not great in several ways, but they could serve as a good starting point.

**Reply by robotlearning123** [reply]
Yes, thank you. I have tried the fish & swimmer model with the old fluid model. Maybe there will be a comparison between the old and new fluid models?

Another question, if change the fish fin to a soft fin, do you have any suggestions? The animal simulation in mujoco is so cool, like the dog. Could we create a similar model in underwater, using the tendon or muscle? Thank you.

**Reply by yuvaltassa** [reply]
We intend to publish a paper and extensive documentation of the new fluid model by the end of the year 🤞.

The easiest thing would be to cut the fin into several parts (a bit like the tail of the fish model) with hinge joint between them. 

Or Perhaps you mean adding a flexible skin visualisation like the dog has? That should also be straightforward.
